File: 01_Array_Hashings/05_Top_K_Frequent_Elements/347_Top_K_Frequent_Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int, int> freq;
        for (int n : nums){
            freq[n]++;
        }

        for (const auto& [key, value] : freq){
            std::cout << key << ": " << value << '\n';
        }

        vector<vector<int>> buckets(nums.size() + 1);

        for (const auto& [key, value]: freq) {
            std::cout << "key: " << key << " value: " << value << "\n";
            buckets[value].push_back(key);
            
        }

        vector<int> res;
        for (int i=buckets.size()-1; i > 0; i--){
            std::cout << i;
            for (int n : buckets[i]){
                res.push_back(n);

            }
            if (res.size() == k)
                return res;
        }

        return res;
    }
};

File: 01_Array_Hashings/05_Top_K_Frequent_Elements/347_Top_K_Frequent_Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = defaultdict(int)
        for n in nums: 
            freq[n] += 1
        
        buckets = [[] for _ in range(len(nums) + 1)]
        for key, value in freq.items():
            buckets[value].append(key)
        res = []
        for i in range(len(buckets) -1, 0, -1):
            for n in buckets[i]:
                res.append(n)
            if len(res) == k:
                return res
        return res